### PR TITLE
Handle IBM DB2 connection re-use using *LOCAL

### LINKED ISSUE #1031: db2 driver does not support blank username and password for *LOCAL connections
### Description

A common idiom on IBM i is to use `*LOCAL` without a username and a password to establish a connection running with the same profile as the process. However, the ADOdb Db2 driver mandates that one must be set.

### Environment
- ADOdb version: 5.22.7
- Driver or Module: Db2
- Database type and version: Db2
- PHP version: PHP 8.3.7
- Platform: IBM i 7.5

* [X] I have tested that the problem is reproducible in the [latest release](https://github.com/ADOdb/ADOdb/releases/latest) / [master](https://github.com/ADOdb/ADOdb/tree/master) branch / latest [hotfix](https://github.com/ADOdb/ADOdb/branches/all?query=hotfix) branch  
  ⚠️ Keep only the applicable option(s), i.e. where you actually tested, and remove the others

### Steps to reproduce

```php
$adoDbDir = "vendor/adodb/adodb-php";
require "$adoDbDir/adodb.inc.php";

$db = newAdoConnection("db2");
var_dump($db->connect("*LOCAL", null, null));

$sql = "select * from qcustcdt";
$array = $db->getAll($sql);

print_r($array);
```

It doesn't seem to handle this very well, continuing on even though `connect` returned false and getting to a point where it uses `false` instead of the connection resource.

```
PHP Fatal error:  Uncaught TypeError: db2_set_option(): Argument #1 ($resource) must be of type resource, false given in /home/calvin/adodbtest/vendor/adodb/adodb-php/drivers/adodb-db2.inc.php:1591
Stack trace:
#0 /home/calvin/adodbtest/vendor/adodb/adodb-php/drivers/adodb-db2.inc.php(1591): db2_set_option()
#1 /home/calvin/adodbtest/vendor/adodb/adodb-php/adodb-lib.inc.php(1236): ADODB_db2->_query()
#2 /home/calvin/adodbtest/vendor/adodb/adodb-php/adodb.inc.php(1611): _adodb_debug_execute()
#3 /home/calvin/adodbtest/vendor/adodb/adodb-php/adodb.inc.php(1590): ADOConnection->_Execute()
#4 /home/calvin/adodbtest/vendor/adodb/adodb-php/adodb.inc.php(2294): ADOConnection->Execute()
#5 /home/calvin/adodbtest/vendor/adodb/adodb-php/adodb.inc.php(2087): ADOConnection->GetArray()
#6 /home/calvin/adodbtest/index.php(21): ADOConnection->GetAll()
#7 {main}
  thrown in /home/calvin/adodbtest/vendor/adodb/adodb-php/drivers/adodb-db2.inc.php on line 1591
```

### Expected behavior

That it would connect fine.

### Additional context 

It appears it's because of this block in `unpackParameters`:

```php
		if ($useCataloguedConnection
		&& (!$argDatabasename
		|| !$argUsername
		|| !$argPassword))
		{

			$errorMessage = 'For catalogued connections, provide ';
			$errorMessage.= 'database, username and password';
			$this->_errorMsg = $errorMessage;
			if ($this->debug)
				ADOConnection::outp($errorMessage);
			return null;

		}
```

This was introduced in the rewrite of this driver in 5.21.

### LINKED ISSUE #1032: db2 driver confuses schema name for database name
### Description

When trying to set the database name when connecting, the Db2 driver mixes up the DSN (containing the host information) and the database name. This is obviously invalid, and will fail to connect.

### Environment
- ADOdb version: 5.22.7
- Driver or Module: Db2
- Database type and version: Db2
- PHP version: 8.3.7
- Platform: IBM i 7.5

* [X] I have tested that the problem is reproducible in the [latest release](https://github.com/ADOdb/ADOdb/releases/latest) / [master](https://github.com/ADOdb/ADOdb/tree/master) branch / latest [hotfix](https://github.com/ADOdb/ADOdb/branches/all?query=hotfix) branch  
  ⚠️ Keep only the applicable option(s), i.e. where you actually tested, and remove the others

### Steps to reproduce

```php
$adoDbDir = "vendor/adodb/adodb-php";
require "$adoDbDir/adodb.inc.php";

$db = newAdoConnection("db2");
var_dump($db->connect("*LOCAL", "username", "password", "QIWS"));

$sql = "select * from qcustcdt"; // this table in QIWS
$array = $db->getAll($sql);

print_r($array);

```

Resulting error spew:

```
Warning: Because you provided user,password and database, DSN connection parameters were discarded
Catalogued connection using parameters: DB=QIWS / UID=username / PWD=password
PHP Warning:  db2_connect(): Connection failure userid (username) in /home/calvin/adodbtest/vendor/adodb/adodb-php/drivers/adodb-db2.inc.php on line 167
PHP Stack trace:
PHP   1. {main}() /home/calvin/adodbtest/index.php:0
PHP   2. ADOConnection->Connect($argHostname = '*LOCAL', $argUsername = 'username', $argPassword = 'password', $argDatabaseName = 'QIWS', $forceNew = *uninitialized*) /home/calvin/adodbtest/index.php:18
PHP   3. ADODB_db2->_connect($argDSN = '*LOCAL', $argUsername = 'username', $argPassword = 'password', $argDatabasename = 'QIWS') /home/calvin/adodbtest/vendor/adodb/adodb-php/adodb.inc.php:942
PHP   4. ADODB_db2->doDB2Connect($argDSN = '*LOCAL', $argUsername = 'username', $argPassword = 'password', $argDatabasename = 'QIWS', $persistent = *uninitialized*) /home/calvin/adodbtest/vendor/adodb/adodb-php/drivers/adodb-db2.inc.php:96
PHP   5. db2_connect($database = 'QIWS', $username = 'username', $password = 'password', $options = []) /home/calvin/adodbtest/vendor/adodb/adodb-php/drivers/adodb-db2.inc.php:167
PHP Warning:  db2_connect(): Relational database QIWS not in relational database directory. SQLCODE=-950 in /home/calvin/adodbtest/vendor/adodb/adodb-php/drivers/adodb-db2.inc.php on line 167
PHP Stack trace:
PHP   1. {main}() /home/calvin/adodbtest/index.php:0
PHP   2. ADOConnection->Connect($argHostname = '*LOCAL', $argUsername = 'username', $argPassword = 'password', $argDatabaseName = 'QIWS', $forceNew = *uninitialized*) /home/calvin/adodbtest/index.php:18
PHP   3. ADODB_db2->_connect($argDSN = '*LOCAL', $argUsername = 'username', $argPassword = 'password', $argDatabasename = 'QIWS') /home/calvin/adodbtest/vendor/adodb/adodb-php/adodb.inc.php:942
PHP   4. ADODB_db2->doDB2Connect($argDSN = '*LOCAL', $argUsername = 'username', $argPassword = 'password', $argDatabasename = 'QIWS', $persistent = *uninitialized*) /home/calvin/adodbtest/vendor/adodb/adodb-php/drivers/adodb-db2.inc.php:96
PHP   5. db2_connect($database = 'QIWS', $username = 'username', $password = 'password', $options = []) /home/calvin/adodbtest/vendor/adodb/adodb-php/drivers/adodb-db2.inc.php:167
*LOCAL: Relational database QIWS not in relational database directory. SQLCODE=-950
```

### Expected behavior

That it would connect to the specified system and set the default database.

### Additional context 

The old pre-rewrite version of the driver in 5.20 seemed to use the DSN intact and called `SET SCHEMA` after connecting: https://github.com/ADOdb/ADOdb/blob/4011379abe34845d44612dbe904f435b6cd81d9f/drivers/adodb-db2.inc.php#L72-L100

### PR DESCRIPTION
**Pull request by mnewnham**
Fixes #1031
Fixes #1032

### CODE REVIEW

File: drivers/adodb-db2.inc.php [hunk 1]
```diff
@@ -198,12 +205,28 @@ private function unpackParameters($argDSN, $argUsername, $argPassword, $argDatab
...
+		{
+			$connectionParameters = array(
+				'dsn'=>'*LOCAL',
+				'uid'=>'',
+				'pwd'=>'',
+				'database'=>'',
+				'catalogue'=>true
+			);
+			return $connectionParameters;
+		}
```
Comment (dregad) [inline drivers/adodb-db2.inc.php:229]: Suggest moving the _if *LOCAL_ block to the top of the function, there is no point in setting $connectionParameters and immediately overwriting its value and return.

**Review by dregad — COMMENTED** [review]
Looks good to me, but I cannot test. 

@NattyNarwhal your feedback on this would be appreciated.

**Comment by mnewnham** [comment]
I have pushed an additional change to the branch regarding SET SCHEMA as suggested.

**Comment by NattyNarwhal** [comment]
Looks like with `"*LOCAL", null, null, "QIWS"` it still doesn't get set. Seems unpackParameters doesn't return said database name in this case.

**Comment by mnewnham** [comment]
As an aside here @NattyNarwhal , It's clear that you are the expert on this. I'm just a guy with a developer edition of DB2 and a SAMPLE database. I've haven't used DB2 in production for more than a decade. @dregad and I  make no claims to expertise on most of the databases we support.  This is your opportunity to shape the driver in any direction you want if you would like to engage on a deeper level. Additionally, I don't have the expertise to test properly. Nonetheless, I've pushed an additional change that might fix this.

File: drivers/adodb-db2.inc.php [hunk 2]
```diff
@@ -133,7 +133,9 @@ private function doDB2Connect($argDSN, $argUsername, $argPassword, $argDatabasen
 		if ($this->debug){
 			if (strcmp($argDSN,'*LOCAL') == 0)
 			{
-				$connectMessage = '*LOCAL connection'; 
+				$connectMessage = '*LOCAL connection';
+				$schema = $argDatabasename;
+				$argDatabasename = '';
```
Comment (dregad) [inline drivers/adodb-db2.inc.php:138]: The schema assignment is wrapped in `if ($this->debug)`, don't you want to do this also when not in debug mode ?

**Comment by NattyNarwhal** [comment]
I'm not too familiar with the ADOdb codebase or Db2 LUW (I mostly work with IBM i), so I'm afraid of regressions there. This patch on top of yours seems to make it work for me at least with the simple test case:

```diff
--- vendor/adodb/adodb-php/drivers/adodb-db2.inc.php.old	2024-05-28 17:16:56.000000000 +0000
+++ vendor/adodb/adodb-php/drivers/adodb-db2.inc.php	2024-05-28 17:20:35.000000000 +0000
@@ -134,8 +134,6 @@
 			if (strcmp($argDSN,'*LOCAL') == 0)
 			{
 				$connectMessage = '*LOCAL connection';
-				$schema = $argDatabasename;
-				$argDatabasename = '';
 			}
 			else if ($useCataloguedConnection)
 			{
@@ -172,14 +170,12 @@
 					$db2Options[$k] = $v;
 		}
 
-		$schema = '';
 		if ($useCataloguedConnection)
 		{
 			$this->_connectionID = $db2Function($argDatabasename,
 												$argUsername,
 												$argPassword,
 												$db2Options);
-			$schema = $argDatabasename;
 		}
 		else
 		
@@ -194,8 +190,8 @@
 		if ($this->_connectionID && $this->connectStmt)
 			$this->execute($this->connectStmt);
 
-		if ($this->_connectionID && $schema)
-			$this->execute("SET SCHEMA=$schema");
+		if ($this->_connectionID && $argDatabasename)
+			$this->execute("SET SCHEMA=$argDatabasename");
 		
 		return $this->_connectionID != false;
 
@@ -230,6 +226,7 @@
 		{
 			$connectionParameters['dsn']      = $argDSN;
 			$connectionParameters['database'] = $argDatabasename;
+			$connectionParameters['catalogue'] = false;
 			
 			return $connectionParameters;
 		}
```

----

That said, I'm not sure if special-casing `*LOCAL` is right; I think it should be treated like any other DSN, just allowing no user/password to be passed.

**Comment by dregad** [comment]
@NattyNarwhal apologies for the long silence. The past few months have been extra busy and I had no available time to spend on ADOdb... Getting back to it this week, preparing release of 5.22.8, I'd like to merge this in if possible.

Your patch does not apply cleanly following Mark's latest commit (464c4869b50f1853762f33eed2ef8b196f277a22), but I believe I have resolved the conflict with the commit I just pushed. Please review and confirm I got it right.

> I'm not sure if special-casing *LOCAL is right; I think it should be treated like any other DSN, just allowing no user/password to be passed.

Since you seem to be the only person in ~20 years to face an issue with *LOCAL on DB2, I'll leave that decision up to you. Feel free to propose an alternative fix if you believe it's needed.

As for the risk of regression, it seems acceptable. We can always patch the driver again later if necessary.

Can you please confirm that with the latest changes, this fixes both #1031 and #1032 for you?

**Comment by NattyNarwhal** [comment]
It seems both `$db->connect("*LOCAL", null, null, "QIWS")` and `$db->connect("*LOCAL", "username", "password", "QIWS")` are working for me now with the current version of the PR - I can use passwordless authentication, and I can set the default schema properly.

**Comment by NattyNarwhal** [comment]
FWIW, if you do need an IBM i system to test the Db2 driver with, let me know; I can provision accounts on one.

**Comment by dregad** [comment]
Thanks for the feedback @NattyNarwhal, I'll proceed with merging this PR then.

With regards to testing on IBM i series, your offer is greatly appreciated, but considering our (lack of) test scripts I'm not sure if or how we could truly make use of it.

**Pull request merged**